File: src/lib/filters.ts
import type {
  MetadataDocument,
  SmartList,
  SortDirection,
  TaskCollection,
  TaskFilter,
  TaskItem,
  TaskOrderField,
  TaskOrdering,
} from '../types'

type SmartListToken =
  | { kind: 'and' | 'or' | 'not' | 'lparen' | 'rparen' }
  | { kind: 'term'; value: string }

type SmartListAst =
  | { kind: 'term'; value: string }
  | { kind: 'not'; child: SmartListAst }
  | { kind: 'and' | 'or'; left: SmartListAst; right: SmartListAst }

export const defaultFilter = (): TaskFilter => ({
  query: '',
  statuses: [],
  tagIds: [],
  includeDescendantTags: true,
  collectionIds: [],
  includeDescendantCollections: true,
  datePreset: 'any',
})

export function createDefaultMetadata(accountId: string): MetadataDocument {
  return {
    accountId,
    version: 2,
    tagNodes: [],
    collectionParents: {},
    collectionOrder: [],
    smartListOrder: [],
    taskListOrderings: {},
    taskListShowCompleted: {},
    manualTaskOrder: {},
    updatedAt: new Date().toISOString(),
  }
}

export function defaultTaskListOrdering(): TaskOrdering {
  return {
    mode: 'manual',
    field: 'dueDate',
    direction: 'asc',
  }
}

export function defaultSmartListOrdering(): TaskOrdering {
  return {
    mode: 'property',
    field: 'dueDate',
    direction: 'asc',
  }
}

export function normalizeOrdering(
  ordering: Partial<TaskOrdering> | undefined,
  fallback: TaskOrdering,
): TaskOrdering {
  return {
    mode: ordering?.mode === 'property' ? 'property' : fallback.mode,
    field: isTaskOrderField(ordering?.field) ? ordering.field : fallback.field,
    direction: isSortDirection(ordering?.direction) ? ordering.direction : fallback.direction,
  }
}

export function extractHashtags(...parts: Array<string | undefined>): string[] {
  const matches = new Set<string>()
  const pattern = /(^|[\s(])#([\p{L}\p{N}_-]+)/gu

  parts.forEach((part) => {
    if (!part) {
      return
    }

    let match: RegExpExecArray | null
    while ((match = pattern.exec(part)) !== null) {
      matches.add(`#${match[2].toLowerCase()}`)
    }
  })

  return Array.from(matches).sort()
}

export function serializeSmartListPayload(smartList: Pick<SmartList, 'definition' | 'ordering' | 'showCompleted'>): string {
  return JSON.stringify(
    {
      definition: smartList.definition,
      ordering: smartList.ordering,
      showCompleted: smartList.showCompleted,
    },
    null,
    2,
  )
}

export function parseFilter(value: string): TaskFilter {
  try {
    const parsed = JSON.parse(value) as Partial<TaskFilter>
    return {
      ...defaultFilter(),
      ...parsed,
      statuses: parsed.statuses ?? [],
      tagIds: (parsed.tagIds ?? []).map(normalizeHashtag).filter(Boolean),
      collectionIds: parsed.collectionIds ?? [],
    }
  } catch {
    return defaultFilter()
  }
}

export function parseSmartListPayload(value: string): {
  definition: string
  ordering: TaskOrdering
  showCompleted: boolean
  legacyFilter?: TaskFilter
} {
  try {
    const parsed = JSON.parse(value) as {
      definition?: string
      filter?: Partial<TaskFilter>
      ordering?: Partial<TaskOrdering>
      showCompleted?: boolean
    }

    if (parsed && typeof parsed === 'object') {
      if (typeof parsed.definition === 'string') {
        return {
          definition: parsed.definition,
          ordering: normalizeOrdering(parsed.ordering, defaultSmartListOrdering()),
          showCompleted: parsed.showCompleted === true,
        }
      }

      if ('filter' in parsed) {
        const legacyFilter = parseFilter(JSON.stringify(parsed.filter ?? {}))
        return {
          definition: '',
          legacyFilter,
          ordering: normalizeOrdering(parsed.ordering, defaultSmartListOrdering()),
          showCompleted: parsed.showCompleted === true,
        }
      }
    }
  } catch {
    // Fall through to legacy string handling.
  }

  return {
    definition: value.trim(),
    ordering: defaultSmartListOrdering(),
    showCompleted: false,
  }
}

function isTaskOrderField(value: unknown): value is TaskOrderField {
  return ['dueDate', 'startDate', 'priority', 'title', 'createdAt', 'updatedAt', 'status'].includes(
    String(value),
  )
}

function isSortDirection(value: unknown): value is SortDirection {
  return value === 'asc' || value === 'desc'
}

export function expandTreeIds(ids: string[], tree: Array<{ id: string; parentId?: string }>): Set<string> {
  const scope = new Set(ids)
  let expanded = true

  while (expanded) {
    expanded = false
    tree.forEach((node) => {
      if (node.parentId && scope.has(node.parentId) && !scope.has(node.id)) {
        scope.add(node.id)
        expanded = true
      }
    })
  }

  return scope
}

function tagScope(filter: TaskFilter): Set<string> | undefined {
  if (filter.tagIds.length === 0) {
    return undefined
  }

  return new Set(filter.tagIds.map(normalizeHashtag).filter(Boolean))
}

function collectionScope(
  filter: TaskFilter,
  metadataDoc: MetadataDocument,
  collections: TaskCollection[],
): Set<string> | undefined {
  if (filter.collectionIds.length === 0) {
    return undefined
  }

  const listTree = collections
    .filter((collection) => collection.kind === 'task')
    .map((collection) => ({
      id: collection.id,
      parentId: metadataDoc.collectionParents[collection.id],
    }))

  const allowedCollections = filter.includeDescendantCollections
    ? expandTreeIds(filter.collectionIds, listTree)
    : new Set(filter.collectionIds)

  return new Set(
    collections
      .filter((collection) => collection.kind === 'task' && allowedCollections.has(collection.id))
      .map((collection) => collection.id),
  )
}

function matchesDate(task: TaskItem, filter: TaskFilter): boolean {
  if (filter.datePreset === 'any') {
    return true
  }

  const candidate = task.dueDate ?? task.startDate
  return matchesDateValue(candidate, task.status, filter)
}

function matchesDateValue(
  candidate: string | undefined,
  taskStatus: TaskItem['status'],
  filter: Pick<TaskFilter, 'datePreset' | 'customFrom' | 'customTo'>,
): boolean {
  if (!candidate) {
    return false
  }

  const date = new Date(candidate)
  const today = new Date()
  today.setHours(0, 0, 0, 0)

  if (filter.datePreset === 'overdue') {
    return date < today && taskStatus !== 'completed'
  }

  if (filter.datePreset === 'today') {
    const end = new Date(today)
    end.setHours(23, 59, 59, 999)
    return date >= today && date <= end
  }

  const upcomingWindow = parseUpcomingDays(filter.datePreset)
  if (upcomingWindow !== undefined) {
    const end = new Date(today)
    end.setDate(end.getDate() + upcomingWindow)
    end.setHours(23, 59, 59, 999)
    return date >= today && date <= end
  }

  const from = filter.customFrom ? new Date(filter.customFrom) : undefined
  const to = filter.customTo ? new Date(filter.customTo) : undefined

  if (from && date < from) {
    return false
  }

  if (to) {
    const end = new Date(to)
    end.setHours(23, 59, 59, 999)
    if (date > end) {
      return false
    }
  }

  return true
}

export function taskMatchesFilter(
  task: TaskItem,
  filter: TaskFilter,
  metadataDoc: MetadataDocument,
  collections: TaskCollection[],
): boolean {
  const query = filter.query.trim().toLowerCase()
  if (query) {
    const haystack = `${task.title} ${task.notes}`.toLowerCase()
    if (!haystack.includes(query)) {
      return false
    }
  }

  if (filter.statuses.length > 0 && !filter.statuses.includes(task.status)) {
    return false
  }

  const taskTags = new Set(task.tagIds.map(normalizeHashtag).filter(Boolean))
  const allowedTags = tagScope(filter)
  if (allowedTags && !Array.from(allowedTags).some((tag) => taskTags.has(tag))) {
    return false
  }

  const allowedCollections = collectionScope(filter, metadataDoc, collections)
  if (allowedCollections && !allowedCollections.has(task.collectionId)) {
    return false
  }

  return matchesDate(task, filter)
}

export function getSmartListCount(
  smartList: SmartList,
  tasks: TaskItem[],
  metadataDoc: MetadataDocument,
  collections: TaskCollection[],
): number {
  return tasks.filter((task) => taskMatchesSmartList(task, smartList, metadataDoc, collections)).length
}

function compareStrings(left: string | undefined, right: string | undefined, direction: SortDirection): number {
  const leftValue = left ?? ''
  const rightValue = right ?? ''
  const result = leftValue.localeCompare(rightValue)
  return direction === 'asc' ? result : -result
}

function compareNumbers(left: number | undefined, right: number | undefined, direction: SortDirection): number {
  const leftValue = left ?? 0
  const rightValue = right ?? 0
  const result = leftValue - rightValue
  return direction === 'asc' ? result : -result
}

function compareByOrdering(left: TaskItem, right: TaskItem, ordering: TaskOrdering): number {
  switch (ordering.field) {
    case 'priority':
      return compareNumbers(left.priority, right.priority, ordering.direction)
    case 'title':
      return compareStrings(left.title, right.title, ordering.direction)
    case 'createdAt':
      return compareStrings(left.createdAt, right.createdAt, ordering.direction)
    case 'updatedAt':
      return compareStrings(left.updatedAt, right.updatedAt, ordering.direction)
    case 'status':
      return compareStrings(left.status, right.status, ordering.direction)
    case 'startDate':
      return compareStrings(left.startDate, right.startDate, ordering.direction)
    case 'dueDate':
    default:
      return compareStrings(left.dueDate, right.dueDate, ordering.direction)
  }
}

export function sortTasks(
  tasks: TaskItem[],
  ordering: TaskOrdering,
  manualTaskIds: string[] = [],
): TaskItem[] {
  const openTasks = tasks.filter((task) => task.status !== 'completed')
  const completedTasks = tasks.filter((task) => task.status === 'completed')

  const orderedOpenTasks =
    ordering.mode === 'manual'
      ? sortManualTasks(openTasks, manualTaskIds)
      : [...openTasks].sort((left, right) => compareTaskWithFallback(left, right, ordering))

  const orderedCompletedTasks = [...completedTasks].sort((left, right) =>
    compareTaskWithFallback(left, right, {
      ...ordering,
      mode: 'property',
    }),
  )

  return [...orderedOpenTasks, ...orderedCompletedTasks]
}

function sortManualTasks(tasks: TaskItem[], manualTaskIds: string[]): TaskItem[] {
  const orderIndex = new Map(manualTaskIds.map((taskId, index) => [taskId, index]))

  return [...tasks].sort((left, right) => {
    const leftIndex = orderIndex.get(left.id)
    const rightIndex = orderIndex.get(right.id)

    if (leftIndex !== undefined && rightIndex !== undefined) {
      return leftIndex - rightIndex
    }
    if (leftIndex !== undefined) {
      return -1
    }
    if (rightIndex !== undefined) {
      return 1
    }

    return compareTaskWithFallback(left, right, {
      mode: 'property',
      field: 'createdAt',
      direction: 'asc',
    })
  })
}

function compareTaskWithFallback(left: TaskItem, right: TaskItem, ordering: TaskOrdering): number {
  const comparisons = [
    compareByOrdering(left, right, ordering),
    compareStrings(left.updatedAt, right.updatedAt, 'desc'),
    compareStrings(left.title, right.title, 'asc'),
    compareStrings(left.id, right.id, 'asc'),
  ]

  return comparisons.find((value) => value !== 0) ?? 0
}

function normalizeHashtag(value: string): string {
  const trimmed = value.trim().toLowerCase()
  if (!trimmed) {
    return ''
  }
  return trimmed.startsWith('#') ? trimmed : `#${trimmed}`
}

function tokenizeSmartListDefinition(definition: string): SmartListToken[] {
  const tokens: SmartListToken[] = []
  let index = 0

  while (index < definition.length) {
    const current = definition[index]

    if (/\s/.test(current)) {
      index += 1
      continue
    }

    if (current === '&') {
      tokens.push({ kind: 'and' })
      index += 1
      continue
    }

    if (current === '|') {
      tokens.push({ kind: 'or' })
      index += 1
      continue
    }

    if (current === '!') {
      tokens.push({ kind: 'not' })
      index += 1
      continue
    }

    if (current === '(') {
      tokens.push({ kind: 'lparen' })
      index += 1
      continue
    }

    if (current === ')') {
      tokens.push({ kind: 'rparen' })
      index += 1
      continue
    }

    if (current === '"') {
      let value = ''
      index += 1
      while (index < definition.length && definition[index] !== '"') {
        if (definition[index] === '\\' && index + 1 < definition.length) {
          index += 1
        }
        value += definition[index]
        index += 1
      }
      if (index >= definition.length) {
        throw new Error('Unterminated quoted string in smart list definition.')
      }
      index += 1
      tokens.push({ kind: 'term', value })
      continue
    }

    const colonQuoteMatch = definition.slice(index).match(/^([a-z_]+):"((?:[^"\\]|\\.)*)"/i)
    if (colonQuoteMatch) {
      tokens.push({
        kind: 'term',
        value: `${colonQuoteMatch[1]}:${colonQuoteMatch[2].replace(/\\"/g, '"')}`,
      })
      index += colonQuoteMatch[0].length
      continue
    }

    let end = index
    while (end < definition.length && !/[\s&|!()]/.test(definition[end])) {
      end += 1
    }
    tokens.push({ kind: 'term', value: definition.slice(index, end) })
    index = end
  }

  return tokens
}

function parseSmartListDefinition(definition: string): SmartListAst | undefined {
  const tokens = tokenizeSmartListDefinition(definition.trim())
  if (tokens.length === 0) {
    return undefined
  }

  let index = 0

  function parseOr(): SmartListAst {
    let left = parseAnd()
    while (tokens[index]?.kind === 'or') {
      index += 1
      left = {
        kind: 'or',
        left,
        right: parseAnd(),
      }
    }
    return left
  }

  function parseAnd(): SmartListAst {
    let left = parseUnary()
    while (tokens[index]?.kind === 'and') {
      index += 1
      left = {
        kind: 'and',
        left,
        right: parseUnary(),
      }
    }
    return left
  }

  function parseUnary(): SmartListAst {
    const token = tokens[index]
    if (!token) {
      throw new Error('Unexpected end of smart list definition.')
    }

    if (token.kind === 'not') {
      index += 1
      return {
        kind: 'not',
        child: parseUnary(),
      }
    }

    if (token.kind === 'lparen') {
      index += 1
      const expression = parseOr()
      if (tokens[index]?.kind !== 'rparen') {
        throw new Error('Missing closing parenthesis in smart list definition.')
      }
      index += 1
      return expression
    }

    if (token.kind === 'term') {
      index += 1
      return {
        kind: 'term',
        value: token.value,
      }
    }

    throw new Error('Invalid smart list definition.')
  }

  const ast = parseOr()
  if (index !== tokens.length) {
    throw new Error('Invalid trailing tokens in smart list definition.')
  }

  return ast
}

function fullTextMatch(task: TaskItem, needle: string): boolean {
  const haystack = `${task.title} ${task.notes}`.toLowerCase()
  return haystack.includes(needle.toLowerCase())
}

function taskMatchesStatusAlias(task: TaskItem, value: string): boolean {
  const normalized = value.toLowerCase()
  if (normalized === 'open') {
    return task.status === 'needs-action' || task.status === 'in-process'
  }
  if (normalized === 'completed') {
    return task.status === 'completed'
  }
  if (normalized === 'in-progress') {
    return task.status === 'in-process'
  }
  return task.status === normalized
}

type SmartDateField = 'start' | 'due' | 'either'

function candidateDateForField(task: TaskItem, field: SmartDateField): string | undefined {
  if (field === 'start') {
    return task.startDate
  }
  if (field === 'due') {
    return task.dueDate
  }
  return task.dueDate ?? task.startDate
}

function taskMatchesNamedDate(task: TaskItem, value: string, field: SmartDateField = 'either'): boolean {
  const upcomingWindow = parseUpcomingDays(value)
  const candidate = candidateDateForField(task, field)

  if (upcomingWindow !== undefined) {
    return matchesDateValue(candidate, task.status, {
      ...defaultFilter(),
      datePreset: `next${upcomingWindow}`,
    })
  }

  return matchesDateValue(candidate, task.status, {
    ...defaultFilter(),
    datePreset: value as TaskFilter['datePreset'],
  })
}

function parseUpcomingDays(value: string): number | undefined {
  const match = /^next(\d+)$/i.exec(value.trim())
  if (!match) {
    return undefined
  }

  const days = Number.parseInt(match[1], 10)
  return Number.isFinite(days) && days > 0 ? days : undefined
}

function parseSmartDateTerm(value: string): { field: SmartDateField; preset: string } | undefined {
  const match = /^(start|due|end):(today|overdue|next\d+)$/i.exec(value.trim())
  if (!match) {
    return undefined
  }

  return {
    field: match[1].toLowerCase() === 'start' ? 'start' : 'due',
    preset: match[2].toLowerCase(),
  }
}

function resolveCollectionIdsByName(value: string, collections: TaskCollection[]): string[] {
  const normalized = value.trim().toLowerCase()
  return collections
    .filter((collection) => collection.kind === 'task' && collection.displayName.trim().toLowerCase() === normalized)
    .map((collection) => collection.id)
}

function taskMatchesDefinitionTerm(
  term: string,
  task: TaskItem,
  metadataDoc: MetadataDocument,
  collections: TaskCollection[],
): boolean {
  const normalized = term.trim()
  const taskTags = new Set(task.tagIds.map(normalizeHashtag).filter(Boolean))
  const lowered = normalized.toLowerCase()

  if (!normalized) {
    return true
  }

  if (normalized.startsWith('#')) {
    return taskTags.has(normalizeHashtag(normalized))
  }

  if (/^p[1-4]$/i.test(normalized)) {
    return task.priority === Number.parseInt(normalized.slice(1), 10)
  }

  const smartDateTerm = parseSmartDateTerm(normalized)
  if (smartDateTerm) {
    return taskMatchesNamedDate(task, smartDateTerm.preset, smartDateTerm.field)
  }

  if (lowered === 'today' || lowered === 'overdue' || parseUpcomingDays(lowered) !== undefined) {
    return taskMatchesNamedDate(task, lowered)
  }

  if (lowered.startsWith('status:')) {
    return taskMatchesStatusAlias(task, normalized.slice('status:'.length))
  }

  if (lowered.startsWith('list:')) {
    const matchingCollectionIds = new Set(resolveCollectionIdsByName(normalized.slice('list:'.length), collections))
    return matchingCollectionIds.has(task.collectionId)
  }

  if (lowered.startsWith('subtree:')) {
    const matchingCollectionIds = resolveCollectionIdsByName(normalized.slice('subtree:'.length), collections)
    const allowed = expandTreeIds(
      matchingCollectionIds,
      collections
        .filter((collection) => collection.kind === 'task')
        .map((collection) => ({
          id: collection.id,
          parentId: metadataDoc.collectionParents[collection.id],
        })),
    )
    return allowed.has(task.collectionId)
  }

  return fullTextMatch(task, normalized)
}

function evaluateSmartListAst(
  ast: SmartListAst,
  task: TaskItem,
  metadataDoc: MetadataDocument,
  collections: TaskCollection[],
): boolean {
  if (ast.kind === 'term') {
    return taskMatchesDefinitionTerm(ast.value, task, metadataDoc, collections)
  }

  if (ast.kind === 'not') {
    return !evaluateSmartListAst(ast.child, task, metadataDoc, collections)
  }

  if (ast.kind === 'and') {
    return (
      evaluateSmartListAst(ast.left, task, metadataDoc, collections) &&
      evaluateSmartListAst(ast.right, task, metadataDoc, collections)
    )
  }

  return (
    evaluateSmartListAst(ast.left, task, metadataDoc, collections) ||
    evaluateSmartListAst(ast.right, task, metadataDoc, collections)
  )
}

export function validateSmartListDefinition(definition: string): string | undefined {
  try {
    parseSmartListDefinition(definition)
    return undefined
  } catch (error) {
    return error instanceof Error ? error.message : 'Invalid smart list definition.'
  }
}

export function taskMatchesSmartList(
  task: TaskItem,
  smartList: Pick<SmartList, 'definition' | 'filter'>,
  metadataDoc: MetadataDocument,
  collections: TaskCollection[],
): boolean {
  const definition = smartList.definition.trim()
  if (definition) {
    try {
      const ast = parseSmartListDefinition(definition)
      return ast ? evaluateSmartListAst(ast, task, metadataDoc, collections) : true
    } catch {
      return false
    }
  }

  return taskMatchesFilter(task, smartList.filter, metadataDoc, collections)
}

export function smartListDefinitionFromFilter(
  filter: TaskFilter,
  collections: TaskCollection[],
): string {
  const terms: string[] = []

  if (filter.query.trim()) {
    terms.push(JSON.stringify(filter.query.trim()))
  }

  filter.statuses.forEach((status) => {
    if (status === 'needs-action') {
      terms.push('status:open')
    } else if (status === 'in-process') {
      terms.push('status:in-progress')
    } else {
      terms.push(`status:${status}`)
    }
  })

  filter.tagIds.forEach((tag) => {
    terms.push(normalizeHashtag(tag))
  })

  filter.collectionIds.forEach((collectionId) => {
    const displayName = collections.find((collection) => collection.id === collectionId)?.displayName ?? collectionId
    terms.push(`${filter.includeDescendantCollections ? 'subtree' : 'list'}:${JSON.stringify(displayName)}`)
  })

  if (filter.datePreset !== 'any') {
    if (
      filter.datePreset === 'today' ||
      filter.datePreset === 'overdue' ||
      parseUpcomingDays(filter.datePreset) !== undefined
    ) {
      terms.push(filter.datePreset)
    }
  }

  return terms.join(' & ')
}

File: desktop_python/taskmanager_desktop/filters.py
from __future__ import annotations

import json
import re
from dataclasses import asdict
from datetime import datetime, timedelta

from .models import MetadataDocument, SmartList, SortDirection, TaskCollection, TaskFilter, TaskItem, TaskOrderField, TaskOrdering


def default_filter() -> TaskFilter:
    return TaskFilter()


def default_task_list_ordering() -> TaskOrdering:
    return TaskOrdering(mode="manual", field="dueDate", direction="asc")


def default_smart_list_ordering() -> TaskOrdering:
    return TaskOrdering(mode="property", field="dueDate", direction="asc")


def create_default_metadata(account_id: str) -> MetadataDocument:
    return MetadataDocument(account_id=account_id, updated_at=datetime.utcnow().isoformat())


def extract_hashtags(*parts: str | None) -> list[str]:
    matches: set[str] = set()
    pattern = re.compile(r"(^|[\s(])#([\w-]+)", re.UNICODE)
    for part in parts:
        if not part:
            continue
        for match in pattern.finditer(part):
            matches.add(f"#{match.group(2).lower()}")
    return sorted(matches)


def serialize_smart_list_payload(smart_list: SmartList) -> str:
    return json.dumps(
        {
            "definition": smart_list.definition,
            "ordering": asdict(smart_list.ordering),
            "showCompleted": smart_list.show_completed,
        },
        indent=2,
    )


def parse_smart_list_payload(value: str) -> tuple[str, TaskOrdering, bool]:
    try:
        parsed = json.loads(value)
    except Exception:
        return value.strip(), default_smart_list_ordering(), False
    if not isinstance(parsed, dict):
        return value.strip(), default_smart_list_ordering(), False
    ordering_dict = parsed.get("ordering") if isinstance(parsed.get("ordering"), dict) else {}
    return (
        str(parsed.get("definition", "")).strip(),
        TaskOrdering(
            mode=ordering_dict.get("mode", "property"),
            field=ordering_dict.get("field", "dueDate"),
            direction=ordering_dict.get("direction", "asc"),
        ),
        parsed.get("showCompleted") is True,
    )


def expand_tree_ids(ids: list[str], tree: list[dict[str, str | None]]) -> set[str]:
    scope = set(ids)
    changed = True
    while changed:
        changed = False
        for node in tree:
            parent = node.get("parentId")
            if parent and parent in scope and node["id"] not in scope:
                scope.add(node["id"])
                changed = True
    return scope


def normalize_ordering(ordering: TaskOrdering | dict | None, fallback: TaskOrdering) -> TaskOrdering:
    if isinstance(ordering, TaskOrdering):
        return ordering
    if not ordering:
        return fallback
    return TaskOrdering(
        mode=ordering.get("mode", fallback.mode),
        field=ordering.get("field", fallback.field),
        direction=ordering.get("direction", fallback.direction),
    )


def smart_list_requires_completed_visibility(smart_list: SmartList) -> bool:
    definition = smart_list.definition.lower()
    return bool(re.search(r"\bcompleted:(today|last\d+)\b", definition) or re.search(r"(^|[^!\w-])status:completed(?=$|[)\s&|])", definition) or smart_list.ordering.field == "completedAt")


def _compare_strings(left: str | None, right: str | None, direction: SortDirection) -> int:
    result = (left or "") > (right or "")
    if (left or "") == (right or ""):
        return 0
    value = 1 if result else -1
    return value if direction == "asc" else -value


def _compare_numbers(left: int | None, right: int | None, direction: SortDirection) -> int:
    delta = (left or 0) - (right or 0)
    if delta == 0:
        return 0
    value = 1 if delta > 0 else -1
    return value if direction == "asc" else -value


def _completed_candidate(task: TaskItem) -> str | None:
    return task.completed_at or (task.updated_at if task.status == "completed" else None)


def _ordering_value(task: TaskItem, field: TaskOrderField):
    if field == "completedAt":
        return _completed_candidate(task)
    if field == "priority":
        return task.priority
    if field == "title":
        return task.title
    if field == "createdAt":
        return task.created_at
    if field == "updatedAt":
        return task.updated_at
    if field == "status":
        return task.status
    if field == "startDate":
        return task.start_date
    return task.due_date


def _compare_tasks(left: TaskItem, right: TaskItem, ordering: TaskOrdering) -> int:
    if ordering.field == "priority":
        result = _compare_numbers(left.priority, right.priority, ordering.direction)
    else:
        result = _compare_strings(_ordering_value(left, ordering.field), _ordering_value(right, ordering.field), ordering.direction)
    if result != 0:
        return result
    return _compare_strings(left.id, right.id, "asc")


def sort_tasks(tasks: list[TaskItem], ordering: TaskOrdering, manual_task_ids: list[str] | None = None) -> list[TaskItem]:
    manual_task_ids = manual_task_ids or []
    if ordering.mode == "manual":
        order_index = {task_id: index for index, task_id in enumerate(manual_task_ids)}
        open_tasks = [task for task in tasks if task.status != "completed"]
        completed_tasks = [task for task in tasks if task.status == "completed"]
        ordered_open = sorted(
            open_tasks,
            key=lambda task: (order_index.get(task.id, 10_000_000), task.created_at, task.id),
        )
        ordered_completed = sorted(completed_tasks, key=lambda task: (_ordering_value(task, ordering.field) or "", task.id))
        return ordered_open + ordered_completed

    from functools import cmp_to_key

    return sorted(tasks, key=cmp_to_key(lambda left, right: _compare_tasks(left, right, ordering)))


def _parse_upcoming_days(value: str) -> int | None:
    match = re.fullmatch(r"next(\d+)", value.strip(), re.I)
    return int(match.group(1)) if match else None


def _parse_past_days(value: str) -> int | None:
    match = re.fullmatch(r"last(\d+)", value.strip(), re.I)
    return int(match.group(1)) if match else None


def _today_bounds() -> tuple[datetime, datetime]:
    start = datetime.now().replace(hour=0, minute=0, second=0, microsecond=0)
    end = start + timedelta(days=1) - timedelta(milliseconds=1)
    return start, end


def _parse_dt(candidate: str | None) -> datetime | None:
    if not candidate:
        return None
    try:
        if len(candidate) == 10:
            return datetime.fromisoformat(candidate)
        return datetime.fromisoformat(candidate.replace("Z", "+00:00"))
    except Exception:
        return None


def _matches_date_value(candidate: str | None, task_status: str, preset: str, custom_from: str | None = None, custom_to: str | None = None) -> bool:
    dt = _parse_dt(candidate)
    if not dt:
        return False
    today_start, today_end = _today_bounds()
    if preset == "any":
        return True
    if preset == "overdue":
        return dt < today_start and task_status != "completed"
    if preset == "today":
        return today_start <= dt <= today_end
    upcoming = _parse_upcoming_days(preset)
    if upcoming is not None:
        return today_start <= dt <= today_end + timedelta(days=upcoming)
    if custom_from and dt < (_parse_dt(custom_from) or dt):
        return False
    if custom_to:
        to_dt = _parse_dt(custom_to)
        if to_dt and dt > (to_dt.replace(hour=23, minute=59, second=59, microsecond=999000)):
            return False
    return True


def task_matches_filter(task: TaskItem, filter_value: TaskFilter, metadata_doc: MetadataDocument, collections: list[TaskCollection]) -> bool:
    query = filter_value.query.strip().lower()
    if query and query not in f"{task.title} {task.notes}".lower():
        return False
    if filter_value.statuses and task.status not in filter_value.statuses:
        return False
    if filter_value.tag_ids and not any(tag in task.tag_ids for tag in filter_value.tag_ids):
        return False
    if filter_value.collection_ids:
        tree = [{"id": collection.id, "parentId": metadata_doc.collection_parents.get(collection.id)} for collection in collections if collection.kind == "task"]
        allowed = expand_tree_ids(filter_value.collection_ids, tree) if filter_value.include_descendant_collections else set(filter_value.collection_ids)
        if task.collection_id not in allowed:
            return False
    return _matches_date_value(task.due_date or task.start_date, task.status, filter_value.date_preset, filter_value.custom_from, filter_value.custom_to)


def _resolve_collection_ids_by_name(name: str, collections: list[TaskCollection]) -> list[str]:
    target = name.strip().strip('"').lower()
    return [collection.id for collection in collections if collection.kind == "task" and collection.display_name.lower() == target]


def _task_matches_named_date(task: TaskItem, preset: str, field: str = "either") -> bool:
    if field == "completed":
        candidate = _completed_candidate(task)
        if not candidate:
            return False
        dt = _parse_dt(candidate)
        if not dt:
            return False
        today_start, today_end = _today_bounds()
        if preset == "today":
            return today_start <= dt <= today_end
        past = _parse_past_days(preset)
        if past is not None:
            return today_start - timedelta(days=past - 1) <= dt <= today_end
        return False
    if field == "start":
        candidate = task.start_date
    elif field in ("due", "end"):
        candidate = task.due_date
    else:
        candidate = task.due_date or task.start_date
    return _matches_date_value(candidate, task.status, preset)


def _term_matches(task: TaskItem, term: str, metadata_doc: MetadataDocument, collections: list[TaskCollection]) -> bool:
    normalized = term.strip()
    lowered = normalized.lower()
    if not normalized:
        return True
    if normalized.startswith("#"):
        return normalized.lower() in [tag.lower() for tag in task.tag_ids]
    if re.fullmatch(r"p[1-4]", lowered):
        return task.priority == int(lowered[1:])
    date_match = re.fullmatch(r"(start|due|end):(today|overdue|next\d+)", lowered)
    if date_match:
        return _task_matches_named_date(task, date_match.group(2), date_match.group(1))
    completed_match = re.fullmatch(r"completed:(today|last\d+)", lowered)
    if completed_match:
        return _task_matches_named_date(task, completed_match.group(1), "completed")
    if lowered in {"today", "overdue"} or _parse_upcoming_days(lowered) is not None:
        return _task_matches_named_date(task, lowered)
    if lowered.startswith("status:"):
        alias = lowered.split(":", 1)[1]
        if alias == "open":
            return task.status in {"needs-action", "in-process"}
        if alias == "in-progress":
            return task.status == "in-process"
        return task.status == alias
    if lowered.startswith("list:"):
        return task.collection_id in set(_resolve_collection_ids_by_name(normalized.split(":", 1)[1], collections))
    if lowered.startswith("subtree:"):
        roots = _resolve_collection_ids_by_name(normalized.split(":", 1)[1], collections)
        allowed = expand_tree_ids(
            roots,
            [{"id": collection.id, "parentId": metadata_doc.collection_parents.get(collection.id)} for collection in collections if collection.kind == "task"],
        )
        return task.collection_id in allowed
    return lowered in f"{task.title} {task.notes}".lower()


def _tokenize(definition: str) -> list[str]:
    pattern = re.compile(r'"(?:[^"\\]|\\.)*"|[()&|!]|[^\s()&|!]+')
    return pattern.findall(definition)


def _parse_expression(tokens: list[str], index: int = 0):
    def parse_or(i: int):
        left, i = parse_and(i)
        while i < len(tokens) and tokens[i] == "|":
            right, i = parse_and(i + 1)
            left = ("or", left, right)
        return left, i

    def parse_and(i: int):
        left, i = parse_unary(i)
        while i < len(tokens) and tokens[i] == "&":
            right, i = parse_unary(i + 1)
            left = ("and", left, right)
        return left, i

    def parse_unary(i: int):
        token = tokens[i]
        if token == "!":
            node, next_i = parse_unary(i + 1)
            return ("not", node), next_i
        if token == "(":
            node, next_i = parse_or(i + 1)
            if tokens[next_i] != ")":
                raise ValueError("Missing closing parenthesis in smart list definition.")
            return node, next_i + 1
        return ("term", token.strip('"')), i + 1

    return parse_or(index)


def task_matches_smart_list(task: TaskItem, smart_list: SmartList, metadata_doc: MetadataDocument, collections: list[TaskCollection]) -> bool:
    definition = smart_list.definition.strip()
    if not definition:
        return task_matches_filter(task, smart_list.filter, metadata_doc, collections)
    tokens = _tokenize(definition)
    if not tokens:
        return True
    ast, index = _parse_expression(tokens)
    if index != len(tokens):
        raise ValueError("Invalid smart list definition.")

    def evaluate(node) -> bool:
        kind = node[0]
        if kind == "term":
            return _term_matches(task, node[1], metadata_doc, collections)
        if kind == "not":
            return not evaluate(node[1])
        if kind == "and":
            return evaluate(node[1]) and evaluate(node[2])
        return evaluate(node[1]) or evaluate(node[2])

    return evaluate(ast)


def validate_smart_list_definition(definition: str) -> str | None:
    try:
        tokens = _tokenize(definition.strip())
        if tokens:
            _parse_expression(tokens)
        return None
    except Exception as error:
        return str(error)
